Readme.md
NProgress
=========

Slim progress bars for Ajax'y applications. Inspired by Google, YouTube, and
Medium.

This version has the ability to display a progress message 
next to the spinner (top right corner). 
All relevant functions (start(), set(), inc(), trickle()) have 
an additional property (msg) that can be any (HTML) string. 
You will definitely want to play with the CSS to make it look good 
on your site, but it looks just great on white background 
(has a glow to the text).

This is also a pure JS version of rsantacruz's [NProgress](http://ricostacruz.com/nprogress), 
which comes form mparramont's fork (see link above). 

You do NOT need jQuery or any other JS library for this version.


Installation
------------

Add [nprogress.js] and [nprogress.css] to your project.

```html
<script src='nprogress.js'></script>
<link rel='stylesheet' href='nprogress.css'/>
```

NProgress is available via [bower] and [npm] and [spm].

    $ bower install --save nprogress
    $ npm install --save nprogress

[bower]: http://bower.io/search/?q=nprogress
[npm]: https://www.npmjs.org/package/nprogress
[spm]: http://spmjs.io/package/nprogress

Basic usage
-----------

Simply call `start()` and `done()` to control the progress bar.

~~~ js
NProgress.start();  // or NProgress.start("Uploading data");
NProgress.done();
~~~

### Turbolinks
Ensure you're using Turbolinks 1.3.0+, and use 
this: (explained [here](https://github.com/rstacruz/nprogress/issues/8#issuecomment-23010560))

~~~ js
$(document).on('page:fetch',   function() { NProgress.start(); });
$(document).on('page:change',  function() { NProgress.done(); });
$(document).on('page:restore', function() { NProgress.remove(); });
~~~

### Pjax
Try this: (explained [here](https://github.com/rstacruz/nprogress/issues/22#issuecomment-36540472))

~~~ js
$(document).on('pjax:start', function() { NProgress.start(); });
$(document).on('pjax:end',   function() { NProgress.done();  });
~~~

Ideas
-----

 * Add progress to your Ajax calls! Bind it to the jQuery `ajaxStart` and
 `ajaxStop` events.

 * Make a fancy loading bar even without Turbolinks/Pjax! Bind it to
 `$(document).ready` and `$(window).load`.

Advanced usage
--------------

__Percentages:__ To set a progress percentage, call `.set(n)`, where *n* is a
number between `0..1`.

~~~ js
NProgress.set(0.0);     // Sorta same as .start()
NProgress.set(0.4, "Uploading form data");
NProgress.set(1.0);     // Sorta same as .done()
~~~

__Incrementing:__ To increment the progress bar, just use `.inc()`. This
increments it with a random amount. This will never get to 100%: use it for
every image load (or similar).

~~~ js
NProgress.inc();
NProgress.inc("default", "A little more");
NProgress.inc(0.2, "20% more");
~~~

If you want to increment by a specific value, you can pass that as a parameter:

~~~ js
NProgress.inc(0.2);    // This will get the current status value and adds 0.2 until status is 0.994
~~~

__Force-done:__ By passing `true` to `done()`, it will show the progress bar
even if it's not being shown. (The default behavior is that *.done()* will not
    do anything if *.start()* isn't called)

~~~ js
NProgress.done(true);
~~~

__Get the status value:__ To get the status value, use `.status`. 
The status value is `null` when the progress bar is not being shown, 
otherwise the status value will be the last perunage set via
`.set()` (and `.inc()`)



Configuration
-------------

#### `minimum`
Changes the minimum percentage used upon starting. (default: `0.08`)

~~~ js
NProgress.configure({ minimum: 0.1 });
~~~

#### `template`
You can change the markup using `template`. To keep the progress
bar working, keep an element with `role='bar'` in there. See the [default template]
for reference.

~~~ js
NProgress.configure({
  template: "<div class='....'>...</div>"
});
~~~

#### `easing` and `speed`
Adjust animation settings using *easing* (a CSS easing string)
and *speed* (in ms). (default: `ease` and `200`)

~~~ js
NProgress.configure({ easing: 'ease', speed: 500 });
~~~

#### `trickle`
Turn off the automatic incrementing behavior by setting this to `false`. (default: `true`)

~~~ js
NProgress.configure({ trickle: false });
~~~

#### `trickleRate` and `trickleSpeed`
You can adjust the *trickleRate* (how much to increase per trickle) and
*trickleSpeed* (how often to trickle, in ms).

~~~ js
NProgress.configure({ trickleRate: 0.02, trickleSpeed: 800 });
~~~

#### `showSpinner`
Turn off loading spinner by setting it to false. (default: `true`)

~~~ js
NProgress.configure({ showSpinner: false });
~~~

#### `parent`
specify this to change the parent container. (default: `body`)

~~~ js
NProgress.configure({ parent: '#container' });
~~~


Customization
-------------

Just edit `variables.less` or `nprogress.less` (which are used to produce 
`nprogress.css`) to your liking. Tip: you probably only want to find
and replace occurrences of `#29d`.

The included CSS file is pretty minimal... in fact, feel free to scrap it and
make your own!

You can regenerate the CSS file using the shell script (after having 
installed NPM and ran `npm install`):

```
./build.sh
```


Resources
---------

 * [New UI Pattern: Website Loading Bars](http://www.usabilitypost.com/2013/08/19/new-ui-pattern-website-loading-bars/) (usabilitypost.com)

Support
-------

__Bugs and requests__: submit them through the project's issues tracker.<br>
[![Issues](http://img.shields.io/github/issues/rstacruz/nprogress.svg)]( https://github.com/rstacruz/nprogress/issues )

__Questions__: ask them at StackOverflow with the tag *nprogress*.<br>
[![StackOverflow](http://img.shields.io/badge/stackoverflow-nprogress-brightgreen.svg)]( http://stackoverflow.com/questions/tagged/nprogress )

__Chat__: join us at gitter.im.<br>
[![Chat](http://img.shields.io/badge/gitter-rstacruz / nprogress-brightgreen.svg)]( https://gitter.im/rstacruz/nprogress )

[default template]:
https://github.com/rstacruz/nprogress/blob/master/nprogress.js#L31
[Turbolinks]: https://github.com/rails/turbolinks
[nprogress.js]: http://ricostacruz.com/nprogress/nprogress.js
[nprogress.css]: http://ricostacruz.com/nprogress/nprogress.css

Thanks
------

**NProgress** © 2013-2014, Rico Sta. Cruz. Released under the [MIT License].<br>
Authored and maintained by Rico Sta. Cruz with help from [contributors].

> [ricostacruz.com](http://ricostacruz.com) &nbsp;&middot;&nbsp;
> GitHub [@rstacruz](https://github.com/rstacruz) &nbsp;&middot;&nbsp;
> Twitter [@rstacruz](https://twitter.com/rstacruz)

[MIT License]: http://mit-license.org/
[contributors]: http://github.com/rstacruz/nprogress/contributors

[![Status](https://api.travis-ci.org/rstacruz/nprogress.svg?branch=master)](http://travis-ci.org/rstacruz/nprogress)
[![npm version](https://img.shields.io/npm/v/nprogress.png)](https://npmjs.org/package/nprogress "View this project on npm")
[![spm package](http://spmjs.io/badge/nprogress)](http://spmjs.io/package/nprogress)



Public API
----------


- __`NProgress.version`__: A string containing the current NProgress version.

- __`NProgress.settings`__: Direct access to the NProgress settings; 
  see above for a comprehensive list of the available settings.

- __`NProgress.configure(settings)`__: Update the NProgress settings. 
  Preferably called before you start NProgress, but this is not mandatory.

  Returns the NProgress instance for easy chaining:

  ```
  NProgress.configure({...}).start().set(0.01, 'blabla');
  ```

- __`NProgress.set(n, msg)`__: Sets the progress bar status, where `n` 
  is a number from `0.0` to `1.0`.

  `msg` is an optional string which will be displayed below the 
  progress bar. The message may contain HTML (innerHTML).

- __`NProgress.isStarted()`__: Returns TRUE when the progress bar 
  is displayed and in progress.

- __`NProgress.start(msg)`__: Shows the progress bar. This is the same 
  as setting the status to 0%, except that it doesn't go backwards.

  `msg` is an optional string which will be displayed below the 
  progress bar. The message may contain HTML (innerHTML).

- __`NProgress.done(force, msg)`__: Hides the progress bar.

  This is the *sort of* the same as setting the status to 100%, with 
  the difference being `done()` makes some placebo effect of some 
  realistic motion.

  If `true` is passed in the `force` argument, it will show the 
  progress bar even if its hidden.

  `msg` is an optional string which will be displayed below the 
  progress bar. The message may contain HTML (innerHTML).

- __`NProgress.inc(amount, msg)`__: Increments by a random amount, 
  unless the `amount` argument is a number, in which case 
  that perunage (number between 0 and 1) will be added to 
  the current progress bar status.

  This method will start showing the progress bar if it isn't 
  displayed yet.

  `msg` is an optional string which will be displayed below the 
  progress bar. The message may contain HTML (innerHTML).

- __`NProgress.trickle(msg)`__: Increments by a small 
  random amount (determined by the `NProgress.settings.trickleRate` 
  setting).

  `msg` is an optional string which will be displayed below 
  the progress bar. The message may contain HTML (innerHTML).

- __`NProgress.showSpinner()`__: Shows the spinner 
  independently from the progress bar.

  Note: Whether this method delivers or not is controlled 
  by the `NProgress.settings.showSpinner` setting.

- __`NProgress.hideSpinner()`__: Hides the spinner 
  independently from the progress bar.

  Note: Whether this method delivers or not is controlled 
  by the `NProgress.settings.showSpinner` setting.

- __`NProgress.showBar()`__: Shows the bar independently 
  from the progress bar.

  Note: Whether this method delivers or not is controlled 
  by the `NProgress.settings.showBar` setting.

- __`NProgress.hideBar()`__: Hides the bar independently 
  from the progress bar.

  Note: Whether this method delivers or not is controlled 
  by the `NProgress.settings.showBar` setting.

- __`NProgress.promise($promise)`__: Waits for all supplied 
  jQuery promises and increases the progress as 
  the promises resolve.

  This method will start showing the progress bar if it 
  isn't displayed yet.

- __`NProgress.isRendered()`__: Checks if the progress bar 
  is rendered.





Internals (API)
---------------

These methods are also available for use by the application code, 
but these are otherwise considered *internal*; while many of these 
help to provide a minimal DOM abstraction layer, a few others are 
very specific to NProgress, e.g. the `.render()` method listed below.


For (un)registering event listeners with NProgress this helper is available: 
you'll only need to call its `addEventListener` and/or `removeEventListener` 
methods when (un)registering your listeners with the 
`NProgress.settings.onDone` and `NProgress.settings.onDoneBefore` events.

- __`NProgress.Internals.generateFunctionRegister()`__: Generator for 
  the new addEventListener/removeEventListener API that sits 
  on top of our events (onDone & onDoneBefore).

  This generator produces a function/class which offers these new 
  addEventListener/removeEventListener methods; calling the generated 
  object directly will invoke all registered handlers.

  Example (pseudo)code:

  ```
  // create new instance of event registrar
  var evtIF = NProgress.Internals.generateFunctionRegister();
  ...
  // register callback(s):
  evtIF.addEventListener(a);
  evtIF.addEventListener(b);
  ...
  // 'fire' the 'event': this will invoke registered callbacks 
  // `a` and `b`, where each will receive the `args` passed 
  // into evtIF:
  evtIF(args);
  ```
  + __`NProgress.Internals.generateFunctionRegister().addEventListener(callback)`__: 
    Add a callback/listener to the event; a la the DOM 
    addEventListener, duplicate registrations are skipped.

    Returns a reference to the current function registrar instance.

  + __`NProgress.Internals.generateFunctionRegister().removeEventListener(callback)`__: 
    Remove the targeted listener callback (if it is still 
    present in the event listener set); when no parameter 
    or a non-function-type parameter is passed, it means 
    *all* listeners will be removed at once.

    Returns a reference to the current function registrar instance.

- __`NProgress.Internals.queue()`__: Returns a queue function/object 
  which queues a function `fn` to be executed. Queued 
  functions are executed with an interval of `NProgress.settings.speed` 
  msecs between them (unless you need them to dequeue 
  faster, which can be accomplished by setting the 
  `.showingProgress` attribute of your `fn` callback/function 
  to a truthy value, in which case this function will 
  dequeue in a 1/20th of the `NProgress.settings.speed` interval.

  + __`NProgress.Internals.queue(fn).next()`__: You can 
    also trigger the dequeue-ing of the next queued 
    function by invoking the `.next()` method directly.

  Example (pseudo)code:

  ```
  // create a new queue instance:
  var q = NProgress.Internals.queue();
  // queue functions to be executed:
  q(f1); q(f2); q(f3);
  q(f4)(f5)(f6);
  ...
  function f1() {
    ...
    // don't wait for the timeout/interval 
    // to invoke the next queued function:
    q.next();
    // and you can even chain these:
    q.next().next().next();
    // and after this the remaining queued 
    // functions will be dequeued at 
    // the designated interval.
  }
  ...
  // and this function will dequeue 
  // much faster once after the previous 
  // dequeued function:
  function q5() { ...}
  q2.showingProgress = true;
  ```

- __`NProgress.Internals.css()`__: Return function which 
  applies CSS properties to an element, similar to 
  the jQuery `css` method, i.e. by auto-detecting if 
  the property is available standard or with a vendor prefix.

  + __`css(element, property, value)`__: apply 
    the value to the given CSS property.

  + __`css(element, properties)`__: apply the 
    values to the given CSS properties listed 
    in the `properties` object.

  Example (pseudo)code:

  ```
  // create a new css instance:
  var css = NProgress.Internals.css();
  ...
  // apply one property value
  css(el, 'display', 'block');
  // remove another property, by setting it to NULL
  css(el, 'opacity', null);
  // apply a set of properties
  css(el, {
    display: 'block',
    position: 'relative'
  });
  ```


Internal NProgress specific methods:

- __`NProgress.render(fromStart)`__: (Internal) renders 
  the progress bar markup based on the `template` setting.

- __`NProgress.remove()`__: (Internal) Removes 
  the progress bar. Opposite of render().

- __`NProgress.getPositioningCSS()`__: Determine 
  which positioning CSS rule to use: sniffs the context 
  and returns one of these, depending on 
  the browser capabilities discovered:

  + 'translate3d': Modern browsers with 3D support, e.g. Webkit, IE10

  + 'translate': Browsers without 3D support, e.g. IE9

  + 'margin': Browsers without translate() support, e.g. IE7-8


Several methods which serve as a very minimal DOM abstraction layer:

- __`NProgress.Internals.hasClass(element, name)`__: Determines 
  if an element or space separated list of class names 
  contains a class name.

- __`NProgress.Internals.addClass(element, name)`__: 
  Adds a class to an element.

- __`NProgress.Internals.removeClass(element, name)`__: 
  Removes a class from an element.

- __`NProgress.Internals.classList(element)`__: 
  Gets a space separated list of the class names on the element.

  The list is wrapped with a single space on each end 
  to facilitate finding matches within the list.

- __`NProgress.Internals.removeElement(element)`__: 
  Removes an element from the DOM.

- __`NProgress.Internals.findElementByAny(root, selector)`__: 
  Return the DOM node for the given `selector`.

  The `selector` can be:

  - **a DOM node** -- in which case this DOM node 
    is immediately produced. This is useful when 
    the user code already has a reference to 
    the desired DOM node and passes it via 
    the NProgress options.

  - **an ID string** -- (without the leading '#')

  - **a query string** -- which identifies a single DOM node. 
    (When it doesn't, the first DOM node that matches 
    will be produced as per
    https://developer.mozilla.org/en-US/docs/Web/API/document.querySelector)



History.md
## [v0.2.0] - May 13, 2015

This is a maintenance release with cleanups. No functionality changes were introduced.

 * [#93] - Remove jQuery dependency in component ([@slang800])
 * [#120] - Fix the Readme showing a wrong example for `ease` ([@teeceepee])

Minor changes:

 * [#92] - Fix typo in Readme ([@FND])
 * [#90] - Fix typo in Readme ([@mdxs])
 * [#119] - Fix typo in Readme ([@infertux])
 * [#53] - Use `===` instead of `==` ([@drodil])
 * Update test dependencies

## [v0.1.6] - June 25, 2014

 * [#86] - Add support for specifying a different parent container. ([@jonjaques], [#14], [#33], [#39])
 * [#64] - Fix Require.js support ([#75], [#85])
 * [#54] - Fix component support

## [v0.1.5] - June 21, 2014

 * [#82] - Release properly to the npm registry

## [v0.1.4] - June 21, 2014

 * [#65] - Bower: don't download package.json. Fixes browserify + jquery problem. ([@amelon])
 * [#75] - Fix compatibility with Karma. ([@shaqq])

Internal changes:

 * [#77] - Use SVG to display Travis-CI badge. ([@Mithgol])
 * [#34] - Readme: update .inc() and .status docs ([@lacivert])
 * [#73] - Readme: update year ([@rwholmes])

## [v0.1.3] - March 26, 2014

 * [#7] - Remove jQuery dependency ([#28], [#17], [@rurjur])
 * [#73] - Update Readme to change year to 2014. ([@rwholmes])

## [v0.1.2] - August 21, 2013

Minor update for proper [Bower] and [Component] support.

 * Add Bower support.
 * Fix Component support and use `component/jquery` as a dependency.

## [v0.1.1] - August 21, 2013

Minor fixes.

 * Removed the busy cursor that occurs when loading.
 * Added support for IE7 to IE9. ([#3], [@markbao])
 * Implement `trickleRate` and `trickleSpeed` options.
 * [#5] - Implement the `showSpinner` option to allow removing the spinner. ([#9], [@rahulcs])
 * Registered as a Component in Component.io.
 * [#8] - Updated the Readme with better Turbolinks instructions.

## v0.1.0 - August 20, 2013

Initial release.

[Bower]: http://bower.io
[Component]: http://component.io
[#119]: https://github.com/rstacruz/nprogress/issues/119
[#120]: https://github.com/rstacruz/nprogress/issues/120
[#14]: https://github.com/rstacruz/nprogress/issues/14
[#17]: https://github.com/rstacruz/nprogress/issues/17
[#28]: https://github.com/rstacruz/nprogress/issues/28
[#33]: https://github.com/rstacruz/nprogress/issues/33
[#34]: https://github.com/rstacruz/nprogress/issues/34
[#39]: https://github.com/rstacruz/nprogress/issues/39
[#3]: https://github.com/rstacruz/nprogress/issues/3
[#54]: https://github.com/rstacruz/nprogress/issues/84
[#5]: https://github.com/rstacruz/nprogress/issues/5
[#64]: https://github.com/rstacruz/nprogress/issues/64
[#65]: https://github.com/rstacruz/nprogress/issues/65
[#73]: https://github.com/rstacruz/nprogress/issues/73
[#75]: https://github.com/rstacruz/nprogress/issues/75
[#77]: https://github.com/rstacruz/nprogress/issues/77
[#7]: https://github.com/rstacruz/nprogress/issues/7
[#82]: https://github.com/rstacruz/nprogress/issues/82
[#84]: https://github.com/rstacruz/nprogress/issues/84
[#85]: https://github.com/rstacruz/nprogress/issues/85
[#86]: https://github.com/rstacruz/nprogress/issues/86
[#8]: https://github.com/rstacruz/nprogress/issues/8
[#90]: https://github.com/rstacruz/nprogress/issues/90
[#92]: https://github.com/rstacruz/nprogress/issues/92
[#93]: https://github.com/rstacruz/nprogress/issues/93
[#9]: https://github.com/rstacruz/nprogress/issues/9
[@slang800]: https://github.com/slang800
[@teeceepee]: https://github.com/teeceepee
[@FND]: https://github.com/FND
[@mdxs]: https://github.com/mdxs
[@infertux]: https://github.com/infertux
[@jonjaques]: https://github.com/jonjaques
[@amelon]: https://github.com/amelon
[@shaqq]: https://github.com/shaqq
[@Mithgol]: https://github.com/Mithgol
[@lacivert]: https://github.com/lacivert
[@rwholmes]: https://github.com/rwholmes
[@rurjur]: https://github.com/rurjur
[@markbao]: https://github.com/markbao
[@rahulcs]: https://github.com/rahulcs
[v0.1.6]: https://github.com/rstacruz/nprogress/compare/v0.1.5...v0.1.6
[v0.1.5]: https://github.com/rstacruz/nprogress/compare/v0.1.4...v0.1.5
[v0.1.4]: https://github.com/rstacruz/nprogress/compare/v0.1.3...v0.1.4
[v0.1.3]: https://github.com/rstacruz/nprogress/compare/v0.1.2...v0.1.3
[v0.1.2]: https://github.com/rstacruz/nprogress/compare/v0.1.1...v0.1.2
[v0.1.1]: https://github.com/rstacruz/nprogress/compare/v0.1.0...v0.1.1
[#53]: https://github.com/rstacruz/nprogress/issues/53
[v0.2.0]: https://github.com/rstacruz/nprogress/compare/v0.1.6...v0.2.0
[@drodil]: https://github.com/drodil


test/test.js
(function() {
  if (typeof process === 'object') {
    require('mocha-jsdom')();
  }

  var root = this;
  var assert = (root.chai || require('chai')).assert;

  describe('NProgress', function() {
    var $, NProgress;

    beforeEach(function() {
      $ = root.jQuery || require('jquery');
      NProgress = root.NProgress || require('../nprogress');

      this.settings = $.extend({}, NProgress.settings);
    });

    afterEach(function() {
      $("#nprogress").remove();
      $('html').attr('class', '');
      NProgress.status = null;

      // Restore settings
      $.extend(NProgress.settings, this.settings);
    });

    describe('.set()', function() {
      it('.set(0) must render', function(done) {
        NProgress.set(0);
        assert.equal($("#nprogress").length, 1);
        assert.equal($("#nprogress .bar").length, 1);
        assert.equal($("#nprogress .peg").length, 1);
        assert.equal($("#nprogress .spinner").length, 1);
        done();
      });

      it('.set(1) should appear and disappear', function(done) {
        NProgress.configure({ speed: 10 });
        NProgress.set(0).set(1);
        assert.equal($("#nprogress").length, 1);

        setTimeout(function() {
          assert.equal($("#nprogress").length, 0);
          done();
        }, 70);
      });

      it('must respect minimum', function() {
        NProgress.set(0);
        assert.equal(NProgress.status, NProgress.settings.minimum);
      });

      it('must clamp to minimum', function() {
        NProgress.set(-100);
        assert.equal(NProgress.status, NProgress.settings.minimum);
      });

      it('must clamp to maximum', function() {
        NProgress.set(456);
        assert.equal(NProgress.status, null);
      });
    });

    // ----

    describe('.start()', function() {
      it('must render', function(done) {
        NProgress.start();
        assert.equal($("#nprogress").length, 1);
        done();
      });

      it('must respect minimum', function() {
        NProgress.start();
        assert.equal(NProgress.status, NProgress.settings.minimum);
      });

      it('must be attached to specified parent', function() {
        var test = $('<div>', {id: 'test'}).appendTo('body');
        NProgress.configure({parent: test});
        NProgress.start();
        assert.isTrue($("#nprogress").parent().is(test));
        assert.isTrue($(NProgress.settings.parent).hasClass("nprogress-parent"));
      });
    });

    // ----

    describe('.done()', function() {
      it('must not render without start', function(done) {
        NProgress.done();
        assert.equal($("#nprogress").length, 0);
        done();
      });

      it('.done(true) must render', function(done) {
        NProgress.done(true);
        assert.equal($("#nprogress").length, 1);
        done();
      });
    });

    // ----

    describe('.remove()', function() {
      it('should be removed from the parent', function() {
        NProgress.set(1);
        NProgress.remove();

        var parent = $(NProgress.settings.parent);
        assert.isFalse(parent.hasClass('nprogress-parent'));
        assert.equal(parent.find('#nprogress').length, 0);
      });
    });

    // ----

    describe('.inc()', function() {
      it('should render', function() {
        NProgress.inc();
        assert.equal($("#nprogress").length, 1);
      });

      it('should start with minimum', function() {
        NProgress.inc();
        assert.equal(NProgress.status, NProgress.settings.minimum);
      });

      it('should increment', function() {
        NProgress.start();
        var start = NProgress.status;

        NProgress.inc();
        assert.operator(NProgress.status, '>', start);
      });

      it('should never reach 1.0', function() {
        for (var i=0; i<100; ++i) { NProgress.inc(); }
        assert.operator(NProgress.status, '<', 1.0);
      });
    });

    // -----

    describe('.configure()', function() {
      it('should work', function() {
        NProgress.configure({ minimum: 0.5 });
        assert.equal(NProgress.settings.minimum, 0.5);
      });
    });

    // ----

    describe('.configure(showSpinner)', function() {
      it('should render spinner by default', function() {
        NProgress.start();

        assert.equal($("#nprogress .spinner").length, 1);
      });

      it('should be true by default', function() {
        assert.equal(NProgress.settings.showSpinner, true);
      });

      it('should hide (on false)', function() {
        NProgress.configure({ showSpinner: false });
        NProgress.start();

        assert.equal($("#nprogress .spinner").length, 0);
      });
    });
  });

})();


guprogress.js
function GUProgress(options) {
	"use strict";
	var self = this;

	// Settings
	self.settings = {
		minimum: 0.08,
		easing: "ease",
		positionUsing: "",
		speed: 200,
		trickle: true,
		trickleRate: 0.02,
		trickleSpeed: 800,
		showSpinner: true,
		template: "<div class=\"guprogress-bar\" role=\"bar\"></div>",
		target: null
	};

	self.status = null;

	self.set = function(n) {
		var started = self.isStarted();

		n = clamp(n, self.settings.minimum, 1);
		self.status = (n === 1 ? null : n);

		var $progress = self.render(!started),
			$bar = $progress.find('[role="bar"]'),
			speed = self.settings.speed,
			ease = self.settings.easing;

		$progress[0].offsetWidth; /* Repaint */

		$progress.queue(function(next) {
			// Set positionUsing if it hasn't already been set
			if (self.settings.positionUsing === '') self.settings.positionUsing = self.getPositioningCSS();

			// Add transition
			$bar.css(barPositionCSS(n, speed, ease));

			if (n === 1) {
				// Fade out
				$progress.css({
					transition: 'none',
					opacity: 1
				});
				$progress[0].offsetWidth; /* Repaint */

				setTimeout(function() {
					$progress.css({
						transition: 'all ' + speed + 'ms linear',
						opacity: 0
					});
					setTimeout(function() {
						self.remove();
						next();
					}, speed);
				}, speed);
			} else {
				setTimeout(next, speed);
			}
		});

		return this;
	};

	self.isStarted = function() {
		return typeof self.status === 'number';
	};

	/**
	 * Shows the progress bar.
	 * This is the same as setting the status to 0%, except that it doesn't go backwards.
	 *
	 *     self.start();
	 *
	 */
	self.start = function() {
		if (!self.status) self.set(0);

		var work = function() {
			setTimeout(function() {
				if (!self.status) return;
				self.trickle();
				work();
			}, self.settings.trickleSpeed);
		};

		if (self.settings.trickle) work();

		return this;
	};

	/**
	 * Hides the progress bar.
	 * This is the *sort of* the same as setting the status to 100%, with the
	 * difference being `done()` makes some placebo effect of some realistic motion.
	 *
	 *     self.done();
	 *
	 * If `true` is passed, it will show the progress bar even if its hidden.
	 *
	 *     self.done(true);
	 */

	self.done = function(force) {
		if (!force && !self.status) return this;

		return self.inc(0.3 + 0.5 * Math.random()).set(1);
	};

	/**
	 * Increments by a random amount.
	 */

	self.inc = function(amount) {
		var n = self.status;

		if (!n) {
			return self.start();
		} else {
			if (typeof amount !== 'number') {
				amount = (1 - n) * clamp(Math.random() * n, 0.1, 0.95);
			}

			n = clamp(n + amount, 0, 0.994);
			return self.set(n);
		}
	};

	self.trickle = function() {
		return self.inc(Math.random() * self.settings.trickleRate);
	};

	/**
	 * Waits for all supplied jQuery promises and
	 * increases the progress as the promises resolve.
	 *
	 * @param $promise jQUery Promise
	 */
	(function() {
		var initial = 0,
			current = 0;

		self.promise = function($promise) {
			if (!$promise || $promise.state() == "resolved") {
				return this;
			}

			if (current == 0) {
				self.start();
			}

			initial++;
			current++;

			$promise.always(function() {
				current--;
				if (current == 0) {
					initial = 0;
					self.done();
				} else {
					self.set((initial - current) / initial);
				}
			});

			return this;
		};

	})();

	/**
	 * (Internal) renders the progress bar markup based on the `template`
	 * setting.
	 */

	self.render = function(fromStart) {
		if (self.isRendered()) return $("#guprogress-self");
		$('html').addClass('self-busy');

		var $el = $("<div id='guprogress-self'>")
			.html(self.settings.template);

		var perc = fromStart ? '-100' : toBarPerc(self.status || 0);

		$el.find('[role="bar"]').css({
			transition: 'all 0 linear',
			transform: 'translate3d(' + perc + '%,0,0)'
		});

		if (!self.settings.showSpinner)
			$el.find('[role="spinner"]').remove();

		$el.prependTo(self.settings.target || document.body);

		return $el;
	};

	/**
	 * Removes the element. Opposite of render().
	 */

	self.remove = function() {
		$('html').removeClass('self-busy');
		$('#guprogress-self').remove();
	};

	/**
	 * Checks if the progress bar is rendered.
	 */

	self.isRendered = function() {
		return ($("#guprogress-self").length > 0);
	};

	/**
	 * Determine which positioning CSS rule to use.
	 */

	self.getPositioningCSS = function() {
		// Sniff on document.body.style
		var target = self.settings.target || document.body;
		var bodyStyle = target.style;

		// Sniff prefixes
		var vendorPrefix = ('WebkitTransform' in bodyStyle) ? 'Webkit' :
			('MozTransform' in bodyStyle) ? 'Moz' :
			('msTransform' in bodyStyle) ? 'ms' :
			('OTransform' in bodyStyle) ? 'O' : '';

		if (vendorPrefix + 'Perspective' in bodyStyle) {
			// Modern browsers with 3D support, e.g. Webkit, IE10
			return 'translate3d';
		} else if (vendorPrefix + 'Transform' in bodyStyle) {
			// Browsers without 3D support, e.g. IE9
			return 'translate';
		} else {
			// Browsers without translate() support, e.g. IE7-8
			return 'margin';
		}
	};

	/**
	 * Helpers
	 */

	function clamp(n, min, max) {
		if (n < min) return min;
		if (n > max) return max;
		return n;
	}

	/**
	 * (Internal) converts a percentage (`0..1`) to a bar translateX
	 * percentage (`-100%..0%`).
	 */

	function toBarPerc(n) {
		return (-1 + n) * 100;
	}


	/**
	 * (Internal) returns the correct CSS for changing the bar's
	 * position given an n percentage, and speed and ease from Settings
	 */

	function barPositionCSS(n, speed, ease) {
		var barCSS;

		if (self.settings.positionUsing === 'translate3d') {
			barCSS = {
				transform: 'translate3d(' + toBarPerc(n) + '%,0,0)'
			};
		} else if (self.settings.positionUsing === 'translate') {
			barCSS = {
				transform: 'translate(' + toBarPerc(n) + '%,0)'
			};
		} else {
			barCSS = {
				'margin-left': toBarPerc(n) + '%'
			};
		}

		barCSS.transition = 'all ' + speed + 'ms ' + ease;

		return barCSS;
	}


	(function init() {
		if (typeof options == "object" && "nodeType" in options && options.nodeType === 1 && options.cloneNode) {
		  	// most probably this is a DOM node, set it as the target
		  	self.settings.target = options;
		} else {
			self.settings = $.extend({}, self.settings, options);
		}
	})();
}